xapian-core/weight/pl2weight.cc

```diff
@@ -70,7 +70,10 @@ PL2Weight::init(double)
     double base_change(1.0 / log(2.0));
     double mean = double(get_collection_freq()) / get_collection_size();
     P1 = mean * base_change + 0.5 * log2(2.0 * M_PI);
-    P2 = log2(mean) + base_change;
+    // If the get_collection_freq() is comparable with get_collection_size()
+    // or mean >= 1/e than P2 will be positive and can lead to a negative
+    // weight, hence take P2 >=0
+    P2 = (mean < (1/2.71828))?(log2(mean) + base_change):0;
```

Comment:
Xapian coding style is to put spaces around operators such as `/`, `?` and `:` for readability.

The "or" in the comment doesn't match the code - the only check in the code is on the value of `mean`.  It's really unhelpful when comments don't match the code they describe - the reader is left in doubt as to whether the comment or code is correct.

But that check doesn't really match the comment (or intent) either.  `e` to the precision of a `double` is not `2.71828`, with the consequence that this actually clamps small positive values of `P2` to 0 as well.

Given this code runs once per weighted term or synonym in the query, the cost of computing the log here is not significant, so it's probably cleanest to simply compare the computed value of `P2` with zero.